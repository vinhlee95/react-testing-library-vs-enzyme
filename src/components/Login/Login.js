import * as React from 'react'

function Login({onSubmit}) {
	const handleSubmit = event => {
		event.preventDefault()
		const {username, password} = event.target.elements
		onSubmit({
			username: username.value,
			password: password.value,
		})
	}

	return (
		<form onSubmit={handleSubmit}>
			<div>
				<label htmlFor="username-field">Username</label>
				<input id="username-field" name="username" type="text" />
			</div>
			<div>
				<label htmlFor="password-field">Password</label>
				<input id="password-field" name="password" type="password" />
			</div>
			<div>
				<button className="submit-button" type="submit" onSubmit={handleSubmit}>
					Submit
				</button>
			</div>
		</form>
	)
}

export default Login
