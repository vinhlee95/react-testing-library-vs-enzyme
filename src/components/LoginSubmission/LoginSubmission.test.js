import * as React from 'react'
import {render, screen, waitForElementToBeRemoved} from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {build, fake} from '@jackfranklin/test-data-bot'

import Login from './LoginSubmission'

const buildLoginForm = build({
	fields: {
		username: fake(f => f.internet.userName()),
		password: fake(f => f.internet.password()),
	},
})

test("logging in displays the user's username", async () => {
	render(<Login />)
	const {username, password} = buildLoginForm()

	userEvent.type(screen.getByLabelText(/username/i), username)
	userEvent.type(screen.getByLabelText(/password/i), password)
	userEvent.click(screen.getByRole('button', {name: /submit/i}))

	await waitForElementToBeRemoved(screen.getByLabelText(/loading/i))
	expect(screen.getByText(username)).toBeInTheDocument()
})

/**
 * Following tests using .toMatchInlineSnapshot()
 * for the benefits of having snapshot values right in the test
 * https://jestjs.io/docs/snapshot-testing#inline-snapshots
 *
 */
describe('logging in without username or password should fail', () => {
	const formData = buildLoginForm()

	it('should fail without username', async () => {
		render(<Login />)
		userEvent.type(screen.getByLabelText(/password/i), formData.password)
		userEvent.click(screen.getByRole('button', {name: /submit/i}))
		await waitForElementToBeRemoved(screen.getByLabelText(/loading/i))
		expect(screen.getByRole('alert')).toMatchInlineSnapshot(`
		<div
		  role="alert"
		  style="color: red;"
		>
		  username is required
		</div>
	`)
	})

	it('should fail without password', async () => {
		render(<Login />)
		userEvent.type(screen.getByLabelText(/username/i), formData.username)
		userEvent.click(screen.getByRole('button', {name: /submit/i}))
		await waitForElementToBeRemoved(screen.getByLabelText(/loading/i))
		expect(screen.getByRole('alert')).toMatchInlineSnapshot(`
		<div
		  role="alert"
		  style="color: red;"
		>
		  password is required
		</div>
	`)
	})
})
