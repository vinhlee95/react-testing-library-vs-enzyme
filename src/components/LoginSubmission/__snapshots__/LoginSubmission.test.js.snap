// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logging in without username or password should fail should fail without password 1`] = `
<div
  role="alert"
  style="color: red;"
>
  password is required
</div>
`;
